jyulb/fileio/isothermal3D/jyu_engine.py

```diff
@@ -0,0 +1,461 @@
+
+
+class JYUEngine(ABCModelingEngine):
+
+    """File-IO wrapper for the JYU-LB Isothermal 3D flow modeling engine.
+
+    Only a single lattice can be added for configuring the simulation;
+    the wrapper does not accept meshes or particle containers.
+
+    The following CUBA keywords are acknowledged in lattice node data:
+    MATERIAL_ID, DENSITY, VELOCITY, and FORCE.
+
+    Some values for the configuration parameters are enumerated: SOLID/FLUID
+    for MATERIAL_ID, STOKES_FLOW/LAMINAR_FLOW/TURBULENT_FLOW for FLOW_TYPE,
+    and BGK/TRT/MRT/REG for COLLISION_OPERATOR.
+
+    Attributes
+    ----------
+    BC : DataContainer
+        container of attributes related to the boundary conditions.
+    CM : DataContainer
+        container of attributes related to the computational method:
+        collision operator, number of time steps, and time step.
+    SP : DataContainer
+        container of attributes related to the system parameters/conditions:
+        kinematic viscosity, reference density, gravity, flow type, and
+        enforcement of external forcing.
+    """
+
+    # Enumeration of some values for the configuration parameters
+    SOLID_ENUM = 0
+    FLUID_ENUM = 255
+
+    STOKES_FLOW_ENUM = 0
+    LAMINAR_FLOW_ENUM = 1
+    TURBULENT_FLOW_ENUM = 2
+
+    BGK_ENUM = 0
+    TRT_ENUM = 1
+    MRT_ENUM = 2
+    REG_ENUM = 3
+
+    def __init__(self):
+        """Initialize and set default parameters for CM, BC, SP, and SD."""
+        # Definition of CM, SP, BC, and SD data components
+        self._data = {}
+        self._lattice = None
+        self.CM = DataContainer()
+        self.SP = DataContainer()
+        self.BC = DataContainer()
+
+        # Default Computational Method data
+        self.CM_CUBA_COLLISION_OPERATOR = JYUEngine.TRT_ENUM
+        self.CM[CUBA.NUMBER_OF_TIME_STEPS] = 10000
+        self.CM[CUBA.TIME_STEP] = 1.0
+        self.base_fname = 'jyu_engine'
+
+        # Default System Parameters data
+        self.SP[CUBA.KINEMATIC_VISCOSITY] = 0.1
+        self.SP_CUBA_REFERENCE_DENSITY = 1.0
+        self.SP_CUBA_GRAVITY = (0.0, 0.0, 0.0)
+        self.SP_CUBA_FLOW_TYPE = JYUEngine.STOKES_FLOW_ENUM
+        self.SP_CUBA_EXTERNAL_FORCING = False
+
+        # Default Boundary Condition data
+        self.BC[CUBA.VELOCITY] = {'open': 'periodic',
+                                  'wall': 'noSlip'}
+
+        self.BC[CUBA.DENSITY] = {'open': 'periodic',
+                                 'wall': 'noFlux'}
+
+    def run(self):
+        """Run the modeling engine using the configured settings.
+
+        Raises
+        ------
+        RuntimeError
+           if a lattice has not been added or
+           if execution of the modeling engine fails.
+        """
+        if self._lattice is None:
+            message = 'A lattice is not added before run in JYUEngine'
+            raise RuntimeError(message)
+
+        # Define input/output file names
+        input_script_fname = self.base_fname + '.in'
```

Comment:
Here I'm just creating several file names with a common body, but different extensions.
That is, no paths involved. It hence appears (from documentations and examples)
that os.path.join is not needed.